### PR TITLE
planner: introduce new cost formula for Selection/TableScan/IndexScan

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #35240

Problem Summary: introduce new cost formula for Selection/TableScan/IndexScan

### What is changed and how it works?

Introduce new cost formula for Selection/TableScan/IndexScan.

On cost model ver2:
1. Selection: `rows*cpu-factor` -> `rows*num-filters*cpu-factor`; use a dedicated `cpu-factor` for TiFlash;
2. TableScan: `rows*row-size*scan-factor` -> `rows*log(row-size)*scan-factor`; use a dedicated `scan-factor` for TiFlash;
3. IndexScan: `rows*row-size*scan-factor` -> `rows*log(row-size)*scan-factor`;

I'll add tests after finishing cost model ver2.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/plan_cost.go [hunk 1]
```diff
@@ -343,12 +367,32 @@ func (p *PhysicalTableScan) GetPlanCost(taskType property.TaskType, costFlag uin
...
+		switch taskType {
+		case property.MppTaskType: // use a dedicated scan-factor for TiFlash
+			scanFactor = p.ctx.GetSessionVars().GetTiFlashScanFactor()
+		default: // for TiKV
+			scanFactor = p.ctx.GetSessionVars().GetScanFactor(p.Table)
+			if p.Desc {
+				scanFactor = p.ctx.GetSessionVars().GetDescScanFactor(p.Table)
+			}
+		}
+		rowSize := math.Max(p.getScanRowSize(), 2.0)
```
Comment (xuyifangreeneyes) [inline planner/core/plan_cost.go:390]: Why cannot `rowSize` exceed 2?

File: planner/core/plan_cost.go [hunk 2]
```diff
@@ -343,12 +367,32 @@ func (p *PhysicalTableScan) GetPlanCost(taskType property.TaskType, costFlag uin
...
+		logRowSize := math.Log2(rowSize)
```
Comment (xuyifangreeneyes) [inline planner/core/plan_cost.go:391]: Log2 in the formula is based on experiment results?

File: planner/core/plan_cost.go [hunk 3]
```diff
@@ -343,12 +367,32 @@ func (p *PhysicalTableScan) GetPlanCost(taskType property.TaskType, costFlag uin
...
+		scanFactor := p.ctx.GetSessionVars().GetScanFactor(p.Table)
+		if p.Desc {
+			scanFactor = p.ctx.GetSessionVars().GetDescScanFactor(p.Table)
+		}
+		selfCost = getCardinality(p, costFlag) * p.getScanRowSize() * scanFactor
+	case modelVer2: // scan cost: rows * log2(row-size) * scan-factor
+		var scanFactor float64
+		switch taskType {
+		case property.MppTaskType: // use a dedicated scan-factor for TiFlash
+			scanFactor = p.ctx.GetSessionVars().GetTiFlashScanFactor()
```
Comment (xuyifangreeneyes) [inline planner/core/plan_cost.go:383]: Here we assume that there is little performance difference between `Scan` and `DescScan` on TiFlash.

File: planner/core/plan_cost.go (hunk 1 above)
Comment (qw4990) [reply planner/core/plan_cost.go:390]: To guarantee `logRowSize >= 1`.

File: planner/core/plan_cost.go (hunk 2 above)
Comment (qw4990) [reply planner/core/plan_cost.go:391]: YES, I'll add some comments here.

File: planner/core/plan_cost.go (hunk 3 above)
Comment (qw4990) [reply planner/core/plan_cost.go:383]: Yes, no need to distinguish `Scan` and `DescScan` for `TiFlash` for now. I'll add some comments.

**Comment by qw4990** [comment]
Some comments were added, PTAL @xuyifangreeneyes

**Review by Reminiscent — APPROVED** [review]

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**